libs/components/packages/src/schematics/migrations/update-7/ag-grid-28.schematic.ts

```diff
@@ -0,0 +1,195 @@
+  }
+}
+
+/**
+ * Upgrade to AG Grid 28 and address three breaking changes:
+ *
+ * - No longer support `@ag-grid-community/all-modules` -- just use the package
+ * - Remove `AgGridModule.withComponents` in module import
+ * - Column API renamed getSecondaryColumns / setSecondaryColumns
+ *
+ * Also advise against mixing modules and packages.
+ */
+export default function (): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    let needsPackageUpdated = false;
+    let needsEntPackageUpdated = false;
+    const agGridAllModules = getPackageJsonDependency(
+      tree,
+      `${ANY_MODULE}all-modules`
+    );
+    const agGridDependency = getPackageJsonDependency(tree, AG_GRID);
+    const agGridAllModulesEnt = getPackageJsonDependency(
+      tree,
+      `${ENT_MODULE}all-modules`
+    );
+    const agGridDependencyEnt = getPackageJsonDependency(tree, AG_GRID_ENT);
+    let type = NodeDependencyType.Default;
+    const overwrite = true;
+    const version = UPDATE_TO_VERSION;
+    if (!agGridAllModules && !agGridDependency) {
+      // AG Grid is not installed, so we don't need to do anything.
+      return;
+    }
+
+    // Determine if there are any community packages that need to be updated.
+    if (
+      agGridDependency &&
+      semver.lt(agGridDependency.version, UPDATE_LT_VERSION)
+    ) {
+      type = agGridDependency.type;
+      needsPackageUpdated = true;
+    } else if (!agGridDependency && agGridAllModules) {
+      type = agGridAllModules.type;
+      needsPackageUpdated = true;
+    }
+    if (needsPackageUpdated) {
+      [AG_GRID, 'ag-grid-angular'].forEach((name) => {
+        addPackageJsonDependency(tree, { name, overwrite, type, version });
+      });
+    }
+    if (agGridAllModules) {
+      removePackageJsonDependency(tree, `${ANY_MODULE}all-modules`);
+      removePackageJsonDependency(tree, `${ANY_MODULE}angular`);
+    }
+
+    // Determine if there are any enterprise packages that need to be updated.
+    if (
+      agGridDependencyEnt &&
+      semver.lt(agGridDependencyEnt.version, UPDATE_LT_VERSION)
+    ) {
+      type = agGridDependencyEnt.type;
+      needsEntPackageUpdated = true;
+    } else if (!agGridDependencyEnt && agGridAllModulesEnt) {
+      type = agGridAllModulesEnt.type;
+      needsEntPackageUpdated = true;
+    }
+    if (needsEntPackageUpdated) {
+      addPackageJsonDependency(tree, {
+        name: AG_GRID_ENT,
+        overwrite,
+        type,
+        version,
+      });
+    }
+    if (agGridAllModulesEnt) {
+      removePackageJsonDependency(tree, `${ENT_MODULE}all-modules`);
+    }
+
+    // Install the new packages.
+    if (
+      needsPackageUpdated ||
+      needsEntPackageUpdated ||
+      agGridAllModules ||
+      agGridAllModulesEnt
+    ) {
+      context.addTask(new NodePackageInstallTask());
+    }
+
+    const ignoreList = tree.exists('.gitignore')
+      ? tree.readText('.gitignore')
+      : '';
+    const ig = ignore().add(ignoreList);
```

Comment:
I'd prefer we didn't add a new dependency just for this one schematic. What's the danger in visiting each `.ts` file even if it's listed in the `.gitignore` file?

Reply:
If you're worried about traversing the `node_modules` directory, you can first get the directory you want to visit:
```
tree.getDir('/projects').visit((filePath) => ...);
tree.getDir('/src').visit((filePath) => ...);
```
See: https://github.com/blackbaud/skyux/blob/main/libs/components/packages/src/schematics/rules/replace-typescript-imports.ts#L44-L45